### PR TITLE
CXX-1187 Handle null return values from libmongoc for uri's

### PR DESCRIPTION
**Pull request by aseering**
This is a proposed fix for [CXX-1187](https://jira.mongodb.org/browse/CXX-1187), as discussed on that ticket.

### CODE REVIEW

File: src/mongocxx/test/uri.cpp [hunk 1]
```diff
@@ -41,4 +43,36 @@ TEST_CASE("URI", "[uri]") {
...
+        REQUIRE(u.password() == "");
+        REQUIRE(u.read_concern().acknowledge_string().empty());
+        REQUIRE(!u.read_preference().tags());
+        REQUIRE(!u.read_preference().max_staleness());
+        REQUIRE(u.replica_set() == "");
+        REQUIRE(u.username() == "");
+        REQUIRE(!u.write_concern().nodes());
+        
+        // Values with defaults with a blank URI
+        REQUIRE(u.ssl() == false);
```
Comment (aseering) [inline src/mongocxx/test/uri.cpp]: I don't know that these values are necessarily correct. I have simply verified that (1) they aren't brazenly unreasonable to my untrained eye, and (2) after the fix below, none of these accessor functions segfault :-)

File: src/mongocxx/uri.cpp [hunk 2]
```diff
@@ -26,6 +26,19 @@
...
+
+std::string to_string_null_safe(const char* str)
+{
+    if (str == nullptr) {
+        return std::string();
+    }
+    return str;
+}
+
+}
```
Comment (aseering) [inline src/mongocxx/uri.cpp]: Stylistically, this is not the most elegant approach ever.  But it works.  I'm guessing that this function may be useful for other wrapper classes as well, but I didn't see a more central place to put it.

File: src/mongocxx/test/uri.cpp [hunk 3]
```diff
@@ -67,7 +67,7 @@ TEST_CASE("URI", "[uri]") {
         REQUIRE(u.hosts().size() == 1);        
         REQUIRE(u.hosts()[0].name == "localhost");
         REQUIRE(u.hosts()[0].port == 27017);
-        REQUIRE(u.hosts()[0].family == 0);
```
Comment (aseering) [inline src/mongocxx/test/uri.cpp]: There's presumably a #define'd constant for this in socket.h on Linux/Mac (AF_INET?), but it's elsewhere on Windows...

File: src/mongocxx/test/uri.cpp [hunk 4]
```diff
@@ -41,4 +43,36 @@ TEST_CASE("URI", "[uri]") {
...
+        // Values with defaults with a blank URI
+        REQUIRE(u.ssl() == false);
+        REQUIRE(u.read_concern().acknowledge_level() == mongocxx::read_concern::level::k_server_default);
+        REQUIRE(u.read_preference().mode() == mongocxx::read_preference::read_mode::k_primary);
+        REQUIRE(u.auth_source() == "admin");
+        REQUIRE(u.hosts().size() == 1);        
+        REQUIRE(u.hosts()[0].name == "localhost");
+        REQUIRE(u.hosts()[0].port == 27017);
+        // Don't check 'u.hosts()[0].family'.  Value is platform-dependent.
+        REQUIRE(u.to_string() == "mongodb://localhost:27017");
```
Comment (jrassi) [inline src/mongocxx/test/uri.cpp]: Suggest replacing the right-hand side with `mongocxx::uri::k_default_uri`.

File: src/mongocxx/uri.cpp (hunk 2 above)
Comment (jrassi) [inline src/mongocxx/uri.cpp]: Oops, one last nit: per code style, please replace this line with:

```
}  // namespace
```

File: src/mongocxx/test/uri.cpp [hunk 5]
```diff
@@ -20,6 +20,8 @@
 #include <mongocxx/exception/logic_error.hpp>
 #include <mongocxx/uri.hpp>
 
+#include <bsoncxx/json.hpp>
```
Comment (aseering) [inline src/mongocxx/test/uri.cpp]: One of the changes suggested in one of your comments below drops the need for this header.  So I think I'll just delete this line, if that's ok.

**Comment by aseering** [comment]
Thanks!  I've updated the code per your suggestions.  Let me know if I missed anything, or if you see anything else that would benefit from a clean-up.

**Comment by jrassi** [comment]
Thanks for addressing my feedback.  This LGTM.  We very much appreciate your report of this issue and your submission of another quality pull request!

Merged in [c045cea](https://github.com/mongodb/mongo-cxx-driver/commit/c045cea1e66b3bebcb96c838efdf99dcb40c6086).

**Pull request closed without merging**